cache/cache.go

```diff
@@ -250,24 +256,61 @@ func (c *Cache) watch(ctx context.Context, rev int64) error {
 			clientv3.WithCreatedNotify(),
 		)
 
-		for resp := range watchCh {
-			readyOnce.Do(func() { c.ready.Set() })
-			if err := resp.Err(); err != nil {
-				c.ready.Reset()
-				c.demux.Purge()
-				return err
+		err := c.watchEvents(ctx, watchCh, applyErr, &readyOnce)
+		c.demux.Unregister(storeW)
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *Cache) applyStorage(ctx context.Context, storeW *watcher, applyErr chan<- error) {
+	defer c.waitGroup.Done()
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
Isn't ctx passed here is the same context as c.internalCtx. Recommend to remove ctx from arguments if you plan to have it as method on Cache.